enclave-runner/src/usercalls/mod.rs

```diff
@@ -545,6 +565,37 @@ fn trap_attached_debugger(tcs: usize) {
     }
 }
 
+/// UsercallExtension provides a mechanism for the enclave code interface with an external service via a modified runner.
+/// If the implemenentation of the functions of UsercallExtension reutrn None, then [default functions](../../fortanix_sgx_abi/struct.Usercalls.html#networking)
+/// would be called which inteface the enclave with TCP stream.
+/// An implementation of UsercallExtension, can be registered while [building](../struct.EnclaveBuilder.html#method.usercall_ext) the enclave.
+
+#[allow(unused)]
+pub trait UsercallExtension : 'static + Send + Sync + std::fmt::Debug {
+    /// Allows the enclave to connect to an external service specified by the addr bytes.
+    /// The Runner should determine the service that the enclave is trying to connect to by parsing the addr bytes.
+    /// If `connect_stream` returns None, the default implementation of [connect_stream](../../fortanix_sgx_abi/struct.Usercalls.html#method.connect_stream) would be called.
+    /// Apart from connecting to the external service, the behavior of this function should be same as the default implementation.
+    fn connect_stream(
+        &self,
+        addr: &[u8],
+        local_addr: Option<&mut String>,
+        peer_addr: Option<&mut String>,
+    ) -> IoResult<Option<Box<SyncStream>>> {
+        Ok(None)
+    }
+}
+
+impl <T: UsercallExtension> From<T> for Box<UsercallExtension> {
+    fn from(value : T) -> Box<UsercallExtension> {
+        Box::new(value)
+    }
+}
```

Comment:
Earlier, I had kept this as Box::new(...), mostly since I was not much used to implementing From/Into on user defined structures. 
But after earlier [code review] (https://github.com/fortanix/rust-sgx/pull/93#discussion_r256693092), I added From conversion.
I am do not have a strong inclination for either, just that I feel into, might abstract the box wrapper for trait dispatch from the user who extends the usercalls. 
Please suggest which alternative to use.

Reply:
Ok, but you're public API is still accepting a `Box<UsercallExtension>` https://github.com/fortanix/rust-sgx/pull/93/files#diff-63c5a8420129862da1b63a4f31723ed8R208, no?

I personally would prefer the public API to just take a `Box<UsercallExtension>` rather than a `T` where `T: Into<Box<UsercallExtension>>`, but if Jethro already asked for the public API to be more generic that's fine.

Reply:
Thanks. made the change to accept `T: Into<Box<UsercallExtension>>` as per 
https://github.com/fortanix/rust-sgx/pull/93#discussion_r256693092